src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,765 @@
+        let mut me = Self::new()?;
+        me.xcb_connection = connection;
+
+        let result = unsafe {
+            (XKBXH.xkb_x11_setup_xkb_extension)(
+                connection,
+                1,
+                2,
+                xkbcommon_dl::x11::xkb_x11_setup_xkb_extension_flags::XKB_X11_SETUP_XKB_EXTENSION_NO_FLAGS,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+            )
+        };
+        assert_eq!(result, 1, "Failed to initialize libxkbcommon");
+
+        unsafe { me.init_with_x11_keymap() };
+
+        Ok(me)
+    }
+
+    #[cfg(feature = "wayland")]
+    pub(crate) fn from_rmlvo(rmlvo: RMLVO) -> Result<Self, Error> {
+        fn to_cstring(s: Option<String>) -> Result<Option<CString>, Error> {
+            s.map_or(Ok(None), |s| CString::new(s).map(Option::Some))
+                .map_err(|_| Error::BadNames)
+        }
+
+        let mut state = Self::new()?;
+
+        let rules = to_cstring(rmlvo.rules)?;
+        let model = to_cstring(rmlvo.model)?;
+        let layout = to_cstring(rmlvo.layout)?;
+        let variant = to_cstring(rmlvo.variant)?;
+        let options = to_cstring(rmlvo.options)?;
+
+        let xkb_names = ffi::xkb_rule_names {
+            rules: rules.map_or(ptr::null(), |s| s.as_ptr()),
+            model: model.map_or(ptr::null(), |s| s.as_ptr()),
+            layout: layout.map_or(ptr::null(), |s| s.as_ptr()),
+            variant: variant.map_or(ptr::null(), |s| s.as_ptr()),
+            options: options.map_or(ptr::null(), |s| s.as_ptr()),
+        };
+
+        unsafe {
+            state.init_with_rmlvo(xkb_names)?;
+        }
+
+        state.locked = true;
+        Ok(state)
+    }
+
+    unsafe fn init_compose(&mut self) {
+        let locale = env::var_os("LC_ALL")
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .or_else(|| env::var_os("LC_CTYPE"))
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .or_else(|| env::var_os("LANG"))
+            .and_then(|v| if v.is_empty() { None } else { Some(v) })
+            .unwrap_or_else(|| "C".into());
+        let locale = CString::new(locale.into_vec()).unwrap();
+
+        let compose_table = (XKBCH.xkb_compose_table_new_from_locale)(
+            self.xkb_context,
+            locale.as_ptr(),
+            ffi::xkb_compose_compile_flags::XKB_COMPOSE_COMPILE_NO_FLAGS,
+        );
+
+        if compose_table.is_null() {
+            // init of compose table failed, continue without compose
+            return;
+        }
+
+        let compose_state = (XKBCH.xkb_compose_state_new)(
+            compose_table,
+            ffi::xkb_compose_state_flags::XKB_COMPOSE_STATE_NO_FLAGS,
+        );
+
+        if compose_table.is_null() {
+            // init of compose state failed, continue without compose
```

Comment:
Could we return something `Result` from this function and log when compose fails? Or add more `log` here wrt why `compose` fails? That could actually be hard to debug.